src/lib.rs

```diff
@@ -343,10 +347,88 @@ impl<O, T: ?Sized> OwningRef<O, T> {
 }
 
 /////////////////////////////////////////////////////////////////////////////
+// OwningHandle
+/////////////////////////////////////////////////////////////////////////////
+
+use std::ops::{Deref, DerefMut};
+
+/// `OwningHandle` is a complement to `OwningRef`. Where `OwningRef` allows
+/// consumers to pass around an owned object and a dependent reference,
+/// `OwningHandle` contains an owned object and a dependent _object_.
+///
+/// `OwningHandle` can encapsulate a `RefMut` along with its associated
+/// `RefCell`, or an `RwLockReadGuard` along with its associated `RwLock`.
+/// However, the API is completely generic and there are no restrictions on
+/// what types of owning and dependent objects may be used.
+///
+/// `OwningHandle` is created by passing an owner object (which dereferences
+/// to a stable address) along with a callback which receives a pointer to
+/// that stable location. The callback may then dereference the pointer and
+/// mint a dependent object, with the guarantee that the returned object will
+/// not outlive the referent of the pointer.
+///
+/// This does foist some unsafety onto the callback, which needs an `unsafe`
+/// block to dereference the pointer. It would be almost good enough for
+/// OwningHandle to pass a transmuted &'statc reference to the callback
+/// since the lifetime is infinite as far as the minted handle is concerned.
+/// However, even an `Fn` callback can still allow the reference to escape
+/// via a `StaticMutex` or similar, which technically violates the safety
+/// contract. Some sort of language support in the lifetime system could
+/// make this API a bit nicer.
+pub struct OwningHandle<O, H>
+    where O: StableAddress, H: Deref,
+{
+    handle: H,
+    _owner: O,
+}
+
+impl<O, H> Deref for OwningHandle<O, H>
+    where O: StableAddress, H: Deref,
+{
+    type Target = H::Target;
+    fn deref(&self) -> &H::Target {
```

Comment:
Maybe this could return `&H`? Then the `H: Deref` bound wouldn’t be necessary. Auto-deref is recursive, so you’d still be able to call e.g. `String` methods on `OwningHandle<_, cell::Ref<String>>`.